### PR TITLE
Change to ALPN based filter chain selection.

### PR DESCRIPTION
**Pull request by incfly**
#6384

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -274,13 +274,15 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
...
 			}
 			if listenerOpts.tlsMultiplexed {
+				// If multiplexed is needed, setup to filter chain match, one for "istio" ALPN TLS, one for default.
 				listenerOpts.filterChainOpts = []*filterChainOpts{
 					{
-						httpOpts:          httpOpts,
-						transportProtocol: authn.EnvoyRawBufferMatch,
+						httpOpts: httpOpts,
+						match: &listener.FilterChainMatch{
+							ApplicationProtocols: ALPNInMesh,
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: pass the ALPN protocols as an option in filter chain opts. Don't build match here

Comment (incfly) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: Either is fine to me. I just feel in future we can build filter chain match based on multiple fields, doing fancier filter chain match...
https://github.com/istio/istio/issues/6384#issuecomment-398259867

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 2]
```diff
@@ -274,13 +274,13 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
 				direction:        http_conn.INGRESS,
 			}
 			if listenerOpts.tlsMultiplexed {
+				// If multiplexed is needed, setup to filter chain match, one for "istio" ALPN TLS, one for default.
```
Comment (PiotrSikora) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: This is describing implementation and not the feature. I'd suggest that you replace it with something like: `Allow both: istio mTLS ("istio" ALPN) and legacy clients.`

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 3]
```diff
@@ -274,13 +274,13 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
 				direction:        http_conn.INGRESS,
 			}
 			if listenerOpts.tlsMultiplexed {
```
Comment (PiotrSikora) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: This and `authnPolicy.RequireTLSMultiplexing` are pretty badly named - again, describing implementation and not the feature. Maybe `acceptIstioAndLegacy` or something like that would be better?

**Comment by PiotrSikora** [comment]
Also, it seems that `buildSidecarListenerTLSContext()` will break since we no longer set `match.TransportProtocol` to `EnvoyRawBufferMatch`.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 3 above)
Comment (incfly) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: Done.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 2 above)
Comment (incfly) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: Done.

**Comment by incfly** [comment]
@PiotrSikora  I changed the logic of `buildSidecarListenerTLSContext`, such that when filter chain match is present, make decision based on the ALPN fields.

Also in order to reference ALPN constants, and avoid cycle importing, I put those definitions in a common.go.

**Comment by incfly** [comment]
And yes, the changes are only for HTTP not TCP. That's the only supported protocol we can configure in authn policy, rbac, etc. IIUC, mixer is also a http filter. Supporting beyond HTTP appears to be larger effort which can be addressed separately.

**Comment by incfly** [comment]
Note I also remove the handling of `Authentication.mTLS.allowTLS`. That option is to allow TLS (no client certificate) traffic. But now that will be matched against the second filter chains when using `PERMISSIVE`, so I don't think we need to have two way of doing that.

@diemtvu Please let me know if that's correct.

If so, we can remove that option in API repo later.

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 4]
```diff
@@ -246,42 +248,47 @@ func BuildAuthNFilter(policy *authn.Policy) *http_conn.HttpFilter {
...
 					},
 				},
-				// Same as ListenersALPNProtocols defined in listener. Need to move that constant else where in order to share.
-				AlpnProtocols: []string{"h2", "http/1.1"},
-			},
-			RequireClientCertificate: &types.BoolValue{
-				Value: !(mTLSParams != nil && mTLSParams.AllowTls),
 			},
+			// Same as ListenersALPNProtocols defined in listener. Need to move that constant else where in order to share.
+			AlpnProtocols: []string{"h2", "http/1.1"},
```
Comment (ZackButcher) [inline pilot/pkg/networking/plugin/authn/authentication.go]: Use a constant here? We have constants for H2, H2 + Istio, why not H2 + HTTP1?

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 5]
```diff
@@ -246,42 +248,47 @@ func BuildAuthNFilter(policy *authn.Policy) *http_conn.HttpFilter {
...
+		},
+		RequireClientCertificate: &types.BoolValue{
+			Value: true,
+		},
+	}
+	// We already decide the purpose of the filter chain, ALPN istio traffic or legacy traffic.
+	if match != nil {
```
Comment (ZackButcher) [inline pilot/pkg/networking/plugin/authn/authentication.go]: Keep these conditions at the top please: there's no reason to build the whole structure if we'll just throw it away part of the time.

```go
if match == nil || !reflect.DeepEqual(match.ApplicationProtocols, common.ALPNInMesh) {
    return nil
}
if requireTLS, _ := RequireTLS(authenticationPolicy); !requireTLS {
    return nil
}

return &auth.DownstreamTlsContext{
 // ...
}
```

File: pilot/pkg/networking/common/common.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2018 Istio Authors
```
Comment (ZackButcher) [inline pilot/pkg/networking/common/common.go]: We already have `pkg/networking/util`; utility/common classes are an anti-pattern to being with, I'd prefer we keep them to a minimum. In this case, just fold this code into the `util` package rather than creating a whole new package to hold three constants.

File: pilot/pkg/networking/plugin/authn/authentication.go (hunk 5 above)
Comment (incfly) [reply pilot/pkg/networking/plugin/authn/authentication.go]: These conditions aren't equivalent to what I want to have.

In my implementation, if `match == nil`, we fall backs to check `RequireTLS`, not returning nil directly.

I think it's okay to sacrifice a little redundant work for readability.

File: pilot/pkg/networking/common/common.go (hunk 6 above)
Comment (incfly) [reply pilot/pkg/networking/common/common.go]: SGTM. Will do.

File: pilot/pkg/networking/plugin/authn/authentication.go (hunk 4 above)
Comment (incfly) [reply pilot/pkg/networking/plugin/authn/authentication.go]: @PiotrSikora what's the recommended name for this constant?

File: pilot/pkg/networking/plugin/authn/authentication.go (hunk 5 above)
Comment (ZackButcher) [reply pilot/pkg/networking/plugin/authn/authentication.go]: Ah sorry, then update the condition to do what you want (`if match != nil && !reflect.DeepEqual`).

Either way I think things are more readable if you put the checks first and exit early rather than doing work and throwing it away. Big +1 on removing the nesting though; just rewrite the conditions to exit early and keep things un-nested as they are in this PR.

File: pilot/pkg/networking/plugin/authn/authentication.go (hunk 4 above)
Comment (PiotrSikora) [reply pilot/pkg/networking/plugin/authn/authentication.go]: Perhaps `ALPNHttp`?

**Comment by PiotrSikora** [comment]
@incfly erm, are you saying that have no way to enable mTLS for non-HTTP traffic? cc @diemtvu

**Comment by incfly** [comment]
@PiotrSikora 

I found when
1. there's only one filter chain,
1. It has an empty filter chain match
1. It does not have `tls_inspector`

then mTLS traffic will fail with "upstream reset the connection". And the plain text traffic can go through.

If so, I will adjust the implementation accordingly, so that for `STRICT` mode, we still don't introduce an empty filter chain match.

**Comment by PiotrSikora** [comment]
@incfly yes, empty filter chain match == no restrictions, so the connection will use the available filter chain. That's intended behavior.

However, since the only filter chain has `tls_context` configured, it will require clients to connect using TLS, otherwise it will reject connections at the TLS layer, so your comment about plaintext connections going through seems to be wrong (or needs investigating if that's indeed happening).

**Comment by incfly** [comment]
@PiotrSikora Thanks, yes, I verified and it's because TlsContext is not set (which is wrong).

After more thoughts, I don't think the current interactions between `listener.go` and `authentication plugin` are right or suitable for this coming PR. Here are cases we need to support

1. Authentication Policy exists, having `STRICT` mode,
   * One filter chain, has `TlsContext` set, no `FilterChainMatch` needed.
1. Authentication Policy, exists, having `PERMISSIVE` mode, needs two filter chains,
   * Filter chain for `istio`: has TlsContext set, has `FilterChainMatch.ALPN` set
   * Filter chain for plaintext: empty filter chain match or no `FilterChainMatch`, no `TlsContext` set.
1. Authentication Policy does not exists. We shouldn't set TlsContext for listener.
   * One filter chain, no `TlsContext` or `FilterChainMatch` set.

Currently, we set up filter chain and its match in `listener.go` by consulting `authentication plugin`, and then `authentication plugin` figure out whether TlsContext is needed. This kinda of "ping-pong" logic is hard to follow and get it right.

What I'm thinking: consolidating above logic in a single place in authentication plugin.

```
func (Plugin) SetupFilterChains() []*FilterChainMatch, []*TlsContext {
}
```

We invoke this method once in listener and we're done. The rest work for authentication plugin is only appending filters as before.

I'll implementing as above. Let me know if you think differently, thanks! @rshriram

**Comment by rshriram** [comment]
From my side I am glad that you settled on introducing a filter chain callback. I was thinking of doing the same. I think you have to be a bit careful around the SNI match for external services - which also use filter chains. But those filter chains shouldn’t technically collide with istio chains except when you have both headless services and external services on 443.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 7]
```diff
@@ -273,21 +263,27 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env model.Env
...
+						env.Mesh, env.IstioConfigStore, instance.Service.Hostname, instance.Endpoint.ServicePort)
+					listenerOpts.requireTLSInspector = requireTLSInspector
+					for i, match := range matches {
+						listenerOpts.filterChainOpts = append(listenerOpts.filterChainOpts, &filterChainOpts{
+							httpOpts:   httpOpts,
+							match:      match,
+							tlsContext: tls[i],
+						})
+					}
+					found = true
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: why can't you wrap this entire code in a callback? like OnFilterChains() callback, thats implemented only by Auth? Pass in the listener opts, etc.. and the plugin can return the filterchain opts.

Comment (incfly) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: Sounds like adding a new method to Plugin interface and only implemented in authn plugin for now.

I'll do do that if that's what you refer to.

**Comment by incfly** [comment]
/hold

still WIP.

**Comment by PiotrSikora** [comment]
@incfly @diemtvu what's the holdup here? Are we really going to ship the version that only works for some use cases instead of this in 1.0?

**Comment by incfly** [comment]
@PiotrSikora @diemtvu @rshriram 
PTAL, now all tests are passing and waiting on a LGTM label, thanks!

File: tests/util/kube_utils.go [hunk 8]
```diff
@@ -508,6 +508,7 @@ func PodExec(n, pod, container, command string, muteOutput bool, kubeconfig stri
 	if muteOutput {
 		return ShellSilent("kubectl exec --kubeconfig=%s %s -n %s -c %s -- %s", kubeconfig, pod, n, container, command)
 	}
+	fmt.Printf("kubectl exec --kubeconfig=%s %s -n %s -c %s -- %s \n", kubeconfig, pod, n, container, command)
```
[the remark this answers is not in the document]
Comment (incfly) [reply tests/util/kube_utils.go]: Done. Removed.

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 9]
```diff
@@ -89,6 +89,89 @@ func RequireTLS(policy *authn.Policy, proxyType model.NodeType) (bool, *authn.Mu
...
+		},
+		RequireClientCertificate: &types.BoolValue{
+			Value: true,
+		},
+	}
+	for _, method := range authnPolicy.Peers {
+		switch method.GetParams().(type) {
+		case *authn.PeerAuthenticationMethod_Mtls:
+			if method.GetMtls().GetMode() == authn.MutualTls_STRICT {
+				log.Infof("Allow only istio mutual TLS traffic")
```
[the remark this answers is not in the document]
Comment (incfly) [reply pilot/pkg/networking/plugin/authn/authentication.go]: Done.

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 10]
```diff
@@ -89,6 +89,89 @@ func RequireTLS(policy *authn.Policy, proxyType model.NodeType) (bool, *authn.Mu
...
+					},
+				},
+			},
+			AlpnProtocols: util.ALPNHttp,
+		},
+		RequireClientCertificate: &types.BoolValue{
+			Value: true,
+		},
+	}
+	for _, method := range authnPolicy.Peers {
```
[the remark this answers is not in the document]
Comment (incfly) [reply pilot/pkg/networking/plugin/authn/authentication.go]: Done. Also renamed to `GetMutualTLS`.

I add a TODO for removing proxyType.

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 11]
```diff
@@ -62,27 +62,123 @@ func NewPlugin() plugin.Plugin {
...
+func JwksURIClusterName(hostname string, port *model.Port) string {
+	const clusterPrefix = "jwks."
+	const maxClusterNameLength = 189 - len(clusterPrefix)
+	name := hostname + "|" + port.Name
+	if len(name) > maxClusterNameLength {
+		prefix := name[:maxClusterNameLength-sha1.Size*2]
+		sum := sha1.Sum([]byte(name))
+		name = fmt.Sprintf("%s%x", prefix, sum)
+	}
+	return clusterPrefix + name
```
Comment (rshriram) [inline pilot/pkg/networking/plugin/authn/authentication.go]: ?? Bad merge / copy paste? This seems like old code (using SHAs etc.)

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 12]
```diff
@@ -752,26 +760,27 @@ type httpListenerOpts struct {
...
-	tlsMultiplexed  bool
+	env                 *model.Environment
+	proxy               *model.Proxy
+	proxyInstances      []*model.ServiceInstance
+	ip                  string
+	port                int
+	protocol            model.Protocol
+	bindToPort          bool
+	filterChainOpts     []*filterChainOpts
+	requireTLSInspector bool
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: why do you need this? This seems to be unused

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 13]
```diff
@@ -262,40 +252,58 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env *model.En
...
+			chains := p.OnInboundFilterChains(params)
+			if len(chains) == 0 {
+				continue
+			}
+			if len(allChains) != 0 {
+				log.Errorf("Found two plugin returns non empty filter chains!")
+				allChains = []plugin.FilterChain{}
+				break
+			}
+			allChains = chains
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: I dont understand this logic: It seems you want to make sure only one plugin returns filter chains
and then outside for loop, you are iterating over all the chains which implies you are prepared for multiple filter chains.. So why not remove that check of len(allchains) != 0.. and simply do 
`allChains = append(allChains, chains…)`

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 14]
```diff
@@ -62,27 +62,123 @@ func NewPlugin() plugin.Plugin {
...
+			ValidationContextType: &auth.CommonTlsContext_ValidationContext{
+				ValidationContext: &auth.CertificateValidationContext{
+					TrustedCa: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: model.AuthCertsPath + model.RootCertFilename,
+						},
+					},
+				},
+			},
+			AlpnProtocols: util.ALPNHttp,
```
[the remark this answers is not in the document]
Comment (incfly) [reply pilot/pkg/networking/plugin/authn/authentication.go]: Good point. Actually I copied from existing values where [LDS's TLS sets it to `h1.1/h2`](https://github.com/istio/istio/blame/master/pilot/pkg/networking/plugin/authn/authentication.go#L262).

I looked and found CDS's TLS using ALPN "istio". If the LDS server side is only "h1.1/h2", does that mean negotiation fail and will fall back to no ALPN value set at all? (In that case, my multiplexing test still succeed, then all traffic will fall back to plain text's filter chain, I'll verify that). Should the ALPN values here  be `ALPNInMesh + ALPNHttp`, aka `{istio, h2, http/1.1}`? @PiotrSikora

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 13 above)
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: Yes. Just remove this check and append. If plugins mess up, then we have a bigger problem.

File: pilot/pkg/networking/plugin/authn/authentication.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (incfly) [reply pilot/pkg/networking/plugin/authn/authentication.go]: I checked and even with current alpn values here permissive mode works and the traffic is indeed in tls(I see handhshake metrics increased.). Add a TODO and will confirm with @PiotrSikora  after he's back.
